tests/unit/moduleapi/customauth.tcl

```diff
@@ -0,0 +1,398 @@
+
+        # Validate that every Blocking AUTH command took at least 500000 usec.
+        set stats [cmdstat auth]
+        regexp "usec_per_call=(\[0-9]{1,})\.*," $stats all usec_per_call
+        assert {$usec_per_call >= 500000}
+
+        r config resetstat
+
+        # Case 5 - Non Blocking Success via the second module.
+        assert_equal {OK} [r AUTH foo allow_two]
+
+        # Case 6 - Non Blocking Deny via the second module.
+        catch { r AUTH foo deny_two } e
+        assert_match {*Auth denied by Misc Module*} $e
+        assert_match {*calls=2,*,rejected_calls=0,failed_calls=1} [cmdstat auth]
+
+        r config resetstat
+
+        # Case 7 - All four auth callbacks "Skip" by not explictly allowing or denying.
+        catch { r AUTH foo nomatch } e
+        assert_match {*WRONGPASS*} $e
+        assert_match {*calls=1,*,rejected_calls=0,failed_calls=1} [cmdstat auth]
+        assert_equal {OK} [r AUTH foo pwd]
+
+        # Because we had to attempt all 4 callbacks, validate that the AUTH command took at least
+        # 1000000 usec (each blocking callback takes 500000 usec).
+        set stats [cmdstat auth]
+        regexp "usec_per_call=(\[0-9]{1,})\.*," $stats all usec_per_call
+        assert {$usec_per_call >= 1000000}
+    }
+
+    test {custom auth during blocking custom auth} {
+        r config resetstat
+        r acl setuser foo >pwd on ~* &* +@all
+        set rd [redis_deferring_client]
+        set rd_two [redis_deferring_client]
+
+        # Attempt blocking custom auth. While this ongoing, attempt non blocking custom auth from
+        # moduleone/moduletwo and start another blocking custom auth from another deferring client.
+        $rd AUTH foo block_allow
+        wait_for_blocked_clients_count 1
+        assert_equal {OK} [r AUTH foo allow]
+        assert_equal {OK} [r AUTH foo allow_two]
+        # Validate that the non blocking custom auth cmds finished before any blocking custom auth.
+        set info_clients [r info clients]
+        assert_match "*blocked_clients:1*" $info_clients
+        $rd_two AUTH foo block_allow
+
+        # Validate that all of the AUTH commands succeeded.
+        wait_for_blocked_clients_count 0 500 10
+        $rd flush
+        assert_equal [$rd read] "OK"
+        $rd_two flush
+        assert_equal [$rd_two read] "OK"
+        assert_match {*calls=4,*,rejected_calls=0,failed_calls=0} [cmdstat auth]
+    }
+
+    test {custom auth inside MULTI EXEC} {
+        r config resetstat
+        r acl setuser foo >pwd on ~* &* +@all
+
+        # Validate that non blocking custom auth inside MULTI succeeds.
+        r multi
+        r AUTH foo allow
+        assert_equal {OK} [r exec]
+
+        # Validate that blocking custom auth inside MULTI throws an err.
+        r multi
+        r AUTH foo block_allow
+        catch {r exec} e
+        assert_match {*ERR Blocking module command called from transaction*} $e
```

Comment:
This error message is automatically added from the engine from existing code in the `moduleBlockClient` function if we try blocking a client (in the middle of LUA or MULTI) from a Module.
```
    if (islua || ismulti) {
        c->bstate.module_blocked_handle = NULL;
        addReplyError(c, islua ?
            "Blocking module command called from Lua script" :
            "Blocking module command called from transaction");
    }
```

I could add conditional logic here to say something like,  "Blocking auth command invoked from a transaction" - if the client is in the middle of module authentication. 

But, I don't think we should be expecting clients to see this error message because the Modules should avoid calling `RM_BlockClient*` APIs since they have the ability to detect if the client is in the middle of LUA or MULTI. Because of this, I thought we can just reuse the same / existing error message